### PR TITLE
Clean name that cluster

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Enhancement / new feature

### Description

Includes the a name in the Virtual Cluster configuration.

### Additional Context

We need a programatic way of identifying which cluster is being used. This allows us to:
- share state at a cluster level.
- Improve logging so its easier to identify which clusters are being used

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by robobario** [comment]
Should we validate that the VCs are uniquely named somewhere?

**Comment by SamBarker** [comment]
> Should we validate that the VCs are uniquely named somewhere?

Yeah for operator sanity/error handling. Currently the last (as seen determined by jackson) cluster with the name will win.

File: kroxylicious/example-proxy-config.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ adminHttp:
     prometheus: {}
 virtualClusters:
   demo:
+    name: "Demo Cluster"
```
Comment (k-wall) [inline kroxylicious/example-proxy-config.yml:13]: Why do we need both the key to the map `virtualClusters` and name in the configuration model?  Should the map be changed into a collection  and we then have a business rule that clusters need to have a distinct name?

**Comment by k-wall** [comment]
Thanks for doing this @SamBarker.  This is also going to pave the way to having metrics distinguished virtual cluster name too.

File: kroxylicious-test-tools/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java [hunk 2]
```diff
@@ -286,12 +289,192 @@ public static Stream<Arguments> fluentApiConfigYamlFidelity() {
 
     @ParameterizedTest(name = "{0}")
     @MethodSource
-    public void fluentApiConfigYamlFidelity(String name, Configuration config, String expected) throws Exception {
+    void fluentApiConfigYamlFidelity(String name, Configuration config, String expected) throws Exception {
```
Comment (k-wall) [inline kroxylicious-test-tools/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java:292]: Noted. Thank you. I was under the misapprehension that public was required, so I'm made the same mistake in a few place across the test suite.  I'll raise a separate PR to fix the others.

**Review by k-wall — APPROVED** [review]
Apart from my question above about whether retaining the `virtualClusters` as a map is a valuable, lgtm.

File: kroxylicious/example-proxy-config.yml (hunk 1 above)
Comment (SamBarker) [reply kroxylicious/example-proxy-config.yml:13]: Hmm good question. 

My thoughts were:
- Less invasive
- Backwards compatible 
- Leaves users more flexibility for human readable names 


Of those the the backwards compatibility is probably the strongest, but none are great on their own

File: kroxylicious-test-tools/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-test-tools/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java:292]: I installed the sonarlint plug-in. I've still got mixed feelings, I think it's overly pedantic and I'm not convinced there is a lot of value in changes such as this...

But going forward it makes sense as a default

File: kroxylicious/example-proxy-config.yml (hunk 1 above)
Comment (SamBarker) [reply kroxylicious/example-proxy-config.yml:13]: One other advantage I thinking about it using a map leads to a more natural layout in the config file. With a map clusters are naturally scoped to by the name.

Comment (SamBarker) [reply kroxylicious/example-proxy-config.yml:13]: We discussed this on [slack](https://kroxylicious.slack.com/archives/C04V1K6EAKZ/p1690834640354339) the TL;DR version is
- the natural scoping is good 
- if we enable [STRICT_DUPLICATE_DETECTION](https://fasterxml.github.io/jackson-core/javadoc/2.9/com/fasterxml/jackson/core/JsonParser.Feature.html#STRICT_DUPLICATE_DETECTION) we can detect duplicates. 
- We don't have any immediate need for display names
- Making configuration problems fatal is a nice property and we can still support re-loading config at some future time

Comment (SamBarker) [reply kroxylicious/example-proxy-config.yml:13]: Enabled `STRICT_DUPLICATE_DETECTION` in 0e868ab84160892c8dcfb328713cf993f61d826d 
and dropped `clusterName` as a config property in cb2faf7a9cb407a92dce6130e3ca17e1ef19e8c5

Comment (robobario) [inline kroxylicious/example-proxy-config.yml:13]: `name` is dead now right, and should be removed from the example-proxy-config?

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/Configuration.java [hunk 3]
```diff
@@ -26,4 +26,10 @@ public List<MicrometerDefinition> getMicrometer() {
     public boolean isUseIoUring() {
         return useIoUring();
     }
+
+    public List<io.kroxylicious.proxy.model.VirtualCluster> validClusters() {
```
Comment (robobario) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/config/Configuration.java:30]: Should this still be `validClusters` when it doesn't filter anything?  It sounds like it returns a subset of clusters.

Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/config/Configuration.java:30]: Its a good question 😁 

I added it so I had somewhere post config parsing to validate the cluster names were unique, and I still feel that it represents a valid and sensible hook to have. However I haven't figured out how to include the port de-conflictor here. Maybe I should just back it out entirely as we are moving to blowing up on invalid config.

**Review by robobario — APPROVED** [review]
LGTM

**Review by k-wall — APPROVED** [review]

**Pull request merged**